File: typescript/src/provider/aws.ts
/* Copyright (c) 2025 Voxgig Ltd, MIT License */

import {
  ProviderSpec, Provider, SekretoError, awsparam, flatname, vaultref, unbase64 } from './support'
import { checkaddr } from './addr'
import { sigv4 } from '../Sigv4'
import { fetchjson } from './http'

function awsnow(): string {
  return new Date()
    .toISOString()
    .replace(/[-:]/g, '')
    .replace(/\.\d+Z$/, 'Z')
}

type Awsopts = {
  region?: string
  keyid?: string
  secret?: string
  session?: string
  addr?: string
  prefix?: string
}

/** Region and credentials, from config first and the standard AWS_*
 * environment variables second - those are AWS's own convention, and a
 * pod or CI job that has them set should just work. Missing either is
 * an error: an AWS store with no credentials could not answer. */
function awsauth(opts: Awsopts): {
  region: string
  keyid: string
  secret: string
  session?: string
} {
  const env = process.env

  const region = opts.region || env.AWS_REGION || env.AWS_DEFAULT_REGION || ''
  const keyid = opts.keyid || env.AWS_ACCESS_KEY_ID || ''
  const secret = opts.secret || env.AWS_SECRET_ACCESS_KEY || ''
  const session = opts.session || env.AWS_SESSION_TOKEN || undefined

  if ('' === region) {
    throw new SekretoError('sekreto: aws: no region (set region or AWS_REGION)')
  }
  if ('' === keyid || '' === secret) {
    throw new SekretoError(
      'sekreto: aws: no credentials (set keyid/secret or AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)',
    )
  }

  return { region, keyid, secret, session }
}

/** One signed call to an AWS JSON-1.1 API. */
async function awscall(
  opts: Awsopts,
  service: string,
  target: string,
  payload: any,
): Promise<{ status: number; body: any }> {
  const auth = awsauth(opts)
  // The China partition lives under its own suffix; every other
  // commercial region is plain amazonaws.com.
  const suffix = auth.region.startsWith('cn-') ? '.amazonaws.com.cn' : '.amazonaws.com'
  const addr = opts.addr || 'https://' + service + '.' + auth.region + suffix
  checkaddr(addr)

  const url = addr.replace(/\/$/, '') + '/'
  const body = JSON.stringify(payload)
  const headers: Record<string, string> = {
    'content-type': 'application/x-amz-json-1.1',
    'x-amz-target': target,
  }

  const signed = sigv4({
    method: 'POST',
    url,
    headers,
    body,
    service,
    region: auth.region,
    keyid: auth.keyid,
    secret: auth.secret,
    session: auth.session,
    datetime: awsnow(),
  })

  return fetchjson('POST', url, { ...headers, ...signed }, body)
}

/** Does this AWS error body name one of the not-found types? Those are
 * a miss; every other failure is a store that could not answer. */
function awsmiss(body: any, types: string[]): boolean {
  const errtype = body && 'string' === typeof body.__type ? body.__type : ''
  return types.some((name) => errtype.includes(name))
}

/** AWS Secrets Manager.
 *
 * `api.token` reads the secret named `api` (the vaultref path, so
 * `db.pass.main` reads `db/pass`) and takes the `token` field of its
 * JSON SecretString - the AWS idiom of one JSON map per secret. A
 * SecretString that is not JSON is the value itself, under the
 * conventional field `value`. Requests are SigV4-signed in-tree; see
 * Sigv4.ts. */
export function awssecretsprovider(options?: Awsopts): Provider {
  const opts = options || {}

  return {
    lookup: async (name: string) => {
      const ref = vaultref(name)

      const res = await awscall(opts, 'secretsmanager', 'secretsmanager.GetSecretValue', {
        SecretId: ref.path,
      })

      if (400 === res.status && awsmiss(res.body, ['ResourceNotFoundException'])) {
        return undefined
      }

      if (200 !== res.status) {
        throw new SekretoError('sekreto: aws secretsmanager error: ' + res.status)
      }

      const text = res.body && res.body.SecretString

      if ('string' !== typeof text) {
        // A binary secret has no fields to address; only the conventional
        // `value` field can mean "the bytes themselves".
        const bin = res.body && res.body.SecretBinary
        if ('string' === typeof bin && 'value' === ref.field) {
          const decoded = unbase64(bin)
          if (undefined === decoded) {
            throw new SekretoError('sekreto: aws secretsmanager: undecodable secret')
          }
          return decoded
        }
        return undefined
      }

      let parsed: any
      try {
        parsed = JSON.parse(text)
      } catch (err: any) {
        parsed = undefined
      }

      if (parsed && 'object' === typeof parsed && !Array.isArray(parsed)) {
        const value = parsed[ref.field]
        return undefined === value || null === value ? undefined : String(value)
      }

      // A plain-string secret is the whole value; it has no named fields.
      return 'value' === ref.field ? text : undefined
    },
    // Config only, never the environment: describe() feeds the spec's
    // sources group, which must answer the same everywhere.
    describe: () => 'awssecrets:' + (opts.region || ''),
  }
}

/** AWS SSM Parameter Store.
 *
 * `db.pass.main` reads the parameter `/db/pass/main` (under an optional
 * prefix path), decrypted. Parameter Store carries flat strings, so
 * there is no field indirection. */
export function awsparamsprovider(options?: Awsopts): Provider {
  const opts = options || {}

  return {
    lookup: async (name: string) => {
      const res = await awscall(opts, 'ssm', 'AmazonSSM.GetParameter', {
        Name: awsparam(name, opts.prefix),
        WithDecryption: true,
      })

      if (400 === res.status && awsmiss(res.body, ['ParameterNotFound'])) {
        return undefined
      }

      if (200 !== res.status) {
        throw new SekretoError('sekreto: aws ssm error: ' + res.status)
      }

      const value = res.body && res.body.Parameter && res.body.Parameter.Value
      return undefined === value || null === value ? undefined : String(value)
    },
    describe: () => 'awsparams:' + (opts.region || '') + (opts.prefix || ''),
  }
}

/** GCP Secret Manager.
 *
 * `api.token` reads secret `api_token` (dots flattened to `_`; Secret
 * Manager ids have no hierarchy and reject dots), latest version. The
 * token comes from config, then `GOOGLE_OAUTH_ACCESS_TOKEN`, then the
 * GCE/GKE metadata server - so on Google's own platform no credential
 * configuration is needed at all.
 *
 * The metadata call itself is plain http to a link-local host by
 * platform design; no credential rides on it, so `checkaddr` guards the
 * Secret Manager address instead. */


// Registering at import is what makes this module's presence the only
// thing that decides whether the kind exists in a build.
import { register } from './Registry'

register({
  name: 'awssecrets',
  needs: ['fetch', 'crypto'],
  define: (spec: ProviderSpec) => awssecretsprovider(spec),
})

register({
  name: 'awsparams',
  needs: ['fetch', 'crypto'],
  define: (spec: ProviderSpec) => awsparamsprovider(spec),
})

File: lua/src/sekreto/plugins/aws.lua
-- The aws plugin: Secrets Manager and SSM Parameter Store, two kinds in
-- one module because they share SigV4 request signing and one signed
-- call.
--
-- SIGV4 TRAVELS WITH THIS PLUGIN. It is the crypto edge - HMAC-SHA256 -
-- and the core of no port imports a hash function, so
-- sekreto.plugins.sigv4 and the SHA-256 in sekreto.plugins.crypto behind
-- it are reached from here and from nowhere else in the library.
--
-- A port of typescript/plugins/aws.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')
local crypto = require('sekreto.plugins.crypto')
local sigv4 = require('sekreto.plugins.sigv4')

local fail = err.fail
local trimslash = providers.trimslash
local providerplugin = providers.providerplugin
local first = support.first
local awsnow = support.awsnow
local fetchjson = httpjson.fetchjson

local M = {}

--- Region and credentials, from config first and the standard AWS_*
--- environment variables second. Missing either is an error: an AWS store
--- with no credentials could not answer.
local function awsauth(region, keyid, secret, session)
  local useregion = first(region, os.getenv('AWS_REGION'), os.getenv('AWS_DEFAULT_REGION'))
  local usekeyid = first(keyid, os.getenv('AWS_ACCESS_KEY_ID'))
  local usesecret = first(secret, os.getenv('AWS_SECRET_ACCESS_KEY'))
  local usesession = first(session, os.getenv('AWS_SESSION_TOKEN'))

  if '' == useregion then
    fail('sekreto: aws: no region (set region or AWS_REGION)')
  end
  if '' == usekeyid or '' == usesecret then
    fail('sekreto: aws: no credentials' ..
      ' (set keyid/secret or AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)')
  end

  return {
    region = useregion,
    keyid = usekeyid,
    secret = usesecret,
    session = ('' == usesession) and nil or usesession,
  }
end

M.awsauth = awsauth

--- One signed call to an AWS JSON-1.1 API.
local function awscall(region, keyid, secret, session, useaddr, service, target, payload)
  local auth = awsauth(region, keyid, secret, session)

  -- The China partition lives under its own suffix; every other
  -- commercial region is plain amazonaws.com.
  local suffix = ('cn-' == auth.region:sub(1, 3)) and '.amazonaws.com.cn'
    or '.amazonaws.com'
  local address = first(useaddr, 'https://' .. service .. '.' .. auth.region .. suffix)
  addr.checkaddr(address)

  local url = trimslash(address) .. '/'

  local extras = {
    { 'content-type', 'application/x-amz-json-1.1' },
    { 'x-amz-target', target },
  }

  local signed = sigv4.sigv4({
    method = 'POST',
    url = url,
    service = service,
    region = auth.region,
    keyid = auth.keyid,
    secret = auth.secret,
    datetime = awsnow(),
    headers = extras,
    body = payload,
    session = auth.session,
  })

  local headers = {}
  for _, pair in ipairs(extras) do
    headers[#headers + 1] = pair
  end
  for _, pair in ipairs(signed) do
    headers[#headers + 1] = pair
  end

  return fetchjson('POST', url, headers, payload)
end

--- Does this AWS error body name one of the not-found types? Those are a
--- miss; every other failure is a store that could not answer.
local function awsmiss(body, ...)
  local errtype = json.asstr(json.dig(body, '__type'))
  if nil == errtype then
    return false
  end

  for _, want in ipairs({ ... }) do
    if nil ~= errtype:find(want, 1, true) then
      return true
    end
  end

  return false
end

M.awsmiss = awsmiss

--- AWS Secrets Manager.
---
--- `api.token` reads the secret named `api` (the vaultref path) and takes
--- the `token` field of its JSON SecretString - the AWS idiom of one JSON
--- map per secret. A SecretString that is not JSON is the value itself,
--- under the conventional field `value`.
local function awssecrets(region, keyid, secret, session, useaddr)
  return {
    lookup = function(name_)
      local ref = name.vaultref(name_)

      local res = awscall(
        region, keyid, secret, session, useaddr,
        'secretsmanager', 'secretsmanager.GetSecretValue',
        json.stringify(json.obj({ { 'SecretId', ref.path } }))
      )

      if 400 == res.status and awsmiss(res.body, 'ResourceNotFoundException') then
        return nil
      end

      if 200 ~= res.status then
        fail('sekreto: aws secretsmanager error: ' .. res.status)
      end

      local text = json.asstr(json.dig(res.body, 'SecretString'))

      if nil == text then
        -- A binary secret has no fields to address; only the conventional
        -- `value` field can mean "the bytes themselves".
        local bin = json.asstr(json.dig(res.body, 'SecretBinary'))
        if nil ~= bin and 'value' == ref.field then
          local decoded = crypto.unbase64(bin)
          if nil == decoded then
            fail('sekreto: aws secretsmanager: undecodable secret')
          end
          return decoded
        end
        return nil
      end

      local parsed = json.parse(text)

      if json.isobj(parsed) then
        return json.text(parsed.vals[ref.field])
      end

      -- A plain-string secret is the whole value; it has no named fields.
      if 'value' == ref.field then
        return text
      end

      return nil
    end,
    -- Config only, never the environment: describe() feeds the spec's
    -- sources group, which must answer the same everywhere.
    describe = function()
      return 'awssecrets:' .. (region or '')
    end,
  }
end

--- AWS SSM Parameter Store.
---
--- `db.pass.main` reads the parameter `/db/pass/main` (under an optional
--- prefix path), decrypted. Parameter Store carries flat strings, so
--- there is no field indirection.
local function awsparams(region, keyid, secret, session, useaddr, prefix)
  return {
    lookup = function(name_)
      local payload = json.obj({
        { 'Name', name.awsparam(name_, prefix) },
        { 'WithDecryption', true },
      })

      local res = awscall(
        region, keyid, secret, session, useaddr,
        'ssm', 'AmazonSSM.GetParameter', json.stringify(payload)
      )

      if 400 == res.status and awsmiss(res.body, 'ParameterNotFound') then
        return nil
      end

      if 200 ~= res.status then
        fail('sekreto: aws ssm error: ' .. res.status)
      end

      return json.text(json.dig(res.body, 'Parameter', 'Value'))
    end,
    describe = function()
      return 'awsparams:' .. (region or '') .. (prefix or '')
    end,
  }
end

--- The two kinds, as voxgig/plugin definitions.
M.awssecrets = providerplugin('awssecrets', function(spec)
  return awssecrets(spec.region, spec.keyid, spec.secret, spec.session, spec.addr)
end)

M.awsparams = providerplugin('awsparams', function(spec)
  return awsparams(spec.region, spec.keyid, spec.secret, spec.session,
    spec.addr, spec.prefix)
end)

return M
